### PR TITLE
Store aggregate replication safety info in catalog and use in ORCA

### PR DESCRIPTION
**Pull request by dgkimura**
In GPDB certain aggregates (specifically order agnostic functions) can
be safely executed on replicated slices. The alternative is to execute
the aggregate on a single segment and then broadcast to the others. The
performance impact of broadcast can be significant on large replicated
datasets.

This commit adds an entry in pg_aggregate to distinguish between
aggregates that can and cannot be safely executed on replicated slices.
By default aggregates are not considered safe to execute on replicated
slices. A user can create safe aggregates by explicitly specifying a new
optional parameter 'repsafe' during CREATE AGGREGATE command.

---

Note to reviewers: I broke this pull-request into separate commits to make it easier to review.  However, I would like to merge as 2 separate commits: one for all catalog changes and another for all ORCA and test changes

### CODE REVIEW

**Comment by soumyadeep2007** [comment]
Haven't taken a look at this yet but we should add `pg_dump` support so that this option can get dumped (for upgrades etc) and we should queue up work for gpbackup.

**Comment by dgkimura** [comment]
On Thursday, January 19, 2023 11:33 AM Soumyadeep Chakraborty wrote:
> we should add pg_dump support so that this option can get dumped (for upgrades etc)

Ah, I had tried to remind myself to do that, but alas I still forgot..  I'll add another commit to do the same.  Thanks!

File: src/backend/gporca/libgpopt/src/base/CUtils.cpp [hunk 1]
```diff
@@ -1907,10 +1907,10 @@ CUtils::PexprCountStar(CMemoryPool *mp)
...
 									  GPOS_NEW(mp) CExpressionArray(mp)));
 
-	CScalarAggFunc *popScAggFunc =
-		PopAggFunc(mp, mdid, str, false /*is_distinct*/,
-				   EaggfuncstageGlobal /*eaggfuncstage*/, false /*fSplit*/,
-				   nullptr, EaggfunckindNormal, GPOS_NEW(mp) ULongPtrArray(mp));
+	CScalarAggFunc *popScAggFunc = PopAggFunc(
+		mp, mdid, str, false /*is_distinct*/,
+		EaggfuncstageGlobal /*eaggfuncstage*/, false /*fSplit*/, nullptr,
+		EaggfunckindNormal, GPOS_NEW(mp) ULongPtrArray(mp), false);
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/base/CUtils.cpp:1913]: We're generating a count(*) here, can we mark this as rep safe then?

File: src/include/catalog/pg_aggregate.dat [hunk 2]
```diff
@@ -643,51 +653,42 @@
...
-  aggtranstype => 'internal' },
+  aggfinalfn => 'percentile_cont_timestamp_final', aggtranstype => 'internal' },
 { aggfnoid => 'median(float8,timestamptz)', aggkind => 'o',
   aggnumdirectargs => '1', aggtransfn => 'ordered_set_transition',
   aggfinalfn => 'percentile_cont_timestamptz_final',
   aggtranstype => 'internal' },
 
 # hyperloglog
-{ aggfnoid => 'gp_hyperloglog_accum(anyelement)', aggkind => 'n',
+{ aggfnoid => 'gp_hyperloglog_accum(anyelement)',
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/include/catalog/pg_aggregate.dat:671]: Wow, good eyes!  Yes, the `reformat_dat_file.p` helper function `strip_default_values` should do exactly that.

File: src/include/catalog/pg_aggregate.dat [hunk 3]
```diff
@@ -153,7 +153,7 @@
   aggtranstype => 'int8' },
 { aggfnoid => 'min(int4)', aggtransfn => 'int4smaller',
   aggcombinefn => 'int4smaller', aggsortop => '<(int4,int4)',
-  aggtranstype => 'int4' },
+  aggtranstype => 'int4', aggrepsafeexec => 't' },
 { aggfnoid => 'min(int2)', aggtransfn => 'int2smaller',
   aggcombinefn => 'int2smaller', aggsortop => '<(int2,int2)',
   aggtranstype => 'int2' },
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/include/catalog/pg_aggregate.dat:159]: Okay, I'll add them too. Thanks!

File: src/backend/gporca/libgpopt/src/base/CUtils.cpp (hunk 1 above)
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/base/CUtils.cpp:1913]: Yeah, I think you're right here.  I'll update

File: src/include/catalog/pg_aggregate.dat [hunk 4]
```diff
@@ -56,11 +57,12 @@
...
+  aggtransspace => '48', aggmtranstype => 'internal', aggmtransspace => '48',
+  aggrepsafeexec => 't' },
 { aggfnoid => 'sum(int4)', aggtransfn => 'int4_sum', aggcombinefn => 'int8pl',
   aggmtransfn => 'int4_avg_accum', aggminvtransfn => 'int4_avg_accum_inv',
   aggmfinalfn => 'int2int4_sum', aggtranstype => 'int8',
-  aggmtranstype => '_int8', aggminitval => '{0,0}' },
+  aggmtranstype => '_int8', aggminitval => '{0,0}', aggrepsafeexec => 't' },
 { aggfnoid => 'sum(int2)', aggtransfn => 'int2_sum', aggcombinefn => 'int8pl',
   aggmtransfn => 'int2_avg_accum', aggminvtransfn => 'int2_avg_accum_inv',
   aggmfinalfn => 'int2int4_sum', aggtranstype => 'int8',
```
Comment (chrishajas) [inline src/include/catalog/pg_aggregate.dat:68]: also for int2

**Review by chrishajas — APPROVED** [review]
LGTM. I believe there are other cases too we can add if we want. min/max should be able to be safely executed for any datatype.

**Review by gpopt — COMMENTED** [review]
Should count(), min(), max() be safe to execute too?

**Comment by charliettxx** [comment]
Just curious between function volatility and safe aggfunc. If tranfunc and combinefunc of aggfunc are both labeled as STABLE or IMMUTABLE,  then could we define this aggfunc as a safe agg? Indeed, we need to consider it in planner as well.

**Comment by dgkimura** [comment]
On Thursday, February 16, 2023 7:08 PM Haotian Chen wrote
> Just curious between function volatility and safe aggfunc. If tranfunc and combinefunc of aggfunc are both labeled as STABLE or IMMUTABLE, then could we define this aggfunc as a safe agg?

Does STABLE or IMMUTABLE guarantee consistent order of the data passed into the aggregate function?  I don't think it does, in which case we cannot make that inference.  If an aggregate function produced different output based on the input order, then even though a single segment it may produce consistent results, it may not match another segment where the data in the table may be ordered different.

**Comment by charliettxx** [comment]
> order agnostic
Got it, the definition ‘safety’ is related to order agnosticof in aggregate. Appreciate your explaination.

**Review by charliettxx — APPROVED** [review]

**Comment by dgkimura** [comment]
Merged. Thanks all for the reviews!

**Pull request merged**